trio/_core/_io_windows.py

```diff
@@ -292,130 +405,37 @@ def _afd_helper_handle():
-# gets super confused. For example, if we issue one operation from
-# wait_readable, and another independent operation from wait_writable, then
-# Windows may complete the wait_writable operation when the socket becomes
-# readable.
-#
-# To avoid this, we have to coalesce all the operations on a single socket
-# into one, and when the set of waiters changes we have to throw away the old
-# operation and start a new one.
-@attr.s(slots=True, eq=False)
-class AFDWaiters:
-    read_task = attr.ib(default=None)
-    write_task = attr.ib(default=None)
-    current_op = attr.ib(default=None)
-
-
-# We also need to bundle up all the info for a single op into a standalone
-# object, because we need to keep all these objects alive until the operation
-# finishes, even if we're throwing it away.
-@attr.s(slots=True, eq=False, frozen=True)
-class AFDPollOp:
-    lpOverlapped = attr.ib()
-    poll_info = attr.ib()
-    waiters = attr.ib()
-    afd_group = attr.ib()
-
-
-# The Windows kernel has a weird issue when using AFD handles. If you have N
-# instances of wait_readable/wait_writable registered with a single AFD handle,
-# then cancelling any one of them takes something like O(N**2) time. So if we
-# used just a single AFD handle, then cancellation would quickly become very
-# expensive, e.g. a program with N active sockets would take something like
-# O(N**3) time to unwind after control-C. The solution is to spread our sockets
-# out over multiple AFD handles, so that N doesn't grow too large for any
-# individual handle.
-MAX_AFD_GROUP_SIZE = 500  # at 1000, the cubic scaling is just starting to bite
-
-
-@attr.s(slots=True, eq=False)
-class AFDGroup:
-    size = attr.ib()
-    handle = attr.ib()
-
-
-@attr.s(slots=True, eq=False, frozen=True)
-class _WindowsStatistics:
-    tasks_waiting_read: int = attr.ib()
-    tasks_waiting_write: int = attr.ib()
-    tasks_waiting_overlapped: int = attr.ib()
-    completion_key_monitors: int = attr.ib()
-    backend: Literal["windows"] = attr.ib(init=False, default="windows")
-
-
-# Maximum number of events to dequeue from the completion port on each pass
-# through the run loop. Somewhat arbitrary. Should be large enough to collect
-# a good set of tasks on each loop, but not so large to waste tons of memory.
-# (Each WindowsIOManager holds a buffer whose size is ~32x this number.)
-MAX_EVENTS = 1000
-
-
 @attr.s(frozen=True)
 class CompletionKeyEventInfo:
-    lpOverlapped = attr.ib()
-    dwNumberOfBytesTransferred = attr.ib()
+    lpOverlapped: None = attr.ib()
+    dwNumberOfBytesTransferred: int = attr.ib()
 
 
 class WindowsIOManager:
-    def __init__(self):
+    def __init__(self) -> None:
         # If this method raises an exception, then __del__ could run on a
         # half-initialized object. So we initialize everything that __del__
         # touches to safe values up front, before we do anything that can
         # fail.
         self._iocp = None
```

Comment:
Could also add a `self._is_initialized` variable that's initially set to `False`, and set to `True` after `self._iocp` is initialized.

Reply:
> Since `iocp` can only be `None` if `__init__` fails (so in `__del__`), it might be better to just cast/type-ignore the none assignment, so we avoid asserts everywhere.

FWIW it can also be `None` after `.close` but IDK if the object still exists then.